sympy/polys/rings.py

```diff
@@ -2531,20 +2531,327 @@ def compose(f, x, a=None):
+        such as `R[x,y][z]`. When dividing `f` by `g` with respect to the variable `z`,
+        the pseudo-quotient `q` and pseudo-remainder `r` satisfy `mf = gq + r`,
+        where `deg(r, z) < deg(g, z)` and `m = LC(g, z)^(deg(f, z) - deg(g, z) + 1)`.
+
+        In this function, the pseudo-remainder `r` can be obtained using the `prem` method,
+        the pseudo-quotient `q` can be obtained using the `pquo` method, and
+        the function `pdiv` itself returns a tuple `(q, r)`
+
+
+        Parameters
+        ==========
 
-    def pexquo(f, g):
-        return f.ring.dmp_exquo(f, g)
+        g : PolyElement
+            The polynomial to divide `self` by.
+        x : generator or generator index, optional
+            The main variable of the polynomials and default is first generator.
+
+        Returns
+        =======
+
+        PolyElement
+            The pseudo-division polynomial (tuple of `q` and `r`).
+
+        Raises
+        ======
+
+        ZeroDivisionError : If `g` is the zero polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy.polys import ring, ZZ
+        >>> R, x, y = ring("x, y", ZZ)
+        >>> f = x**2 + x*y
+        >>> g = 2*x + 2
+        >>> f.pdiv(g) # first generator is chosen by default if it is not given
+        (2*x + 2*y - 2, -4*y + 4)
+
+        >>> f.pdiv(g, y) # generator is given
+        (2*x**3 + 2*x**2*y + 6*x**2 + 2*x*y + 8*x + 4, 0)
+        >>> f.pdiv(g, 1) # generator index is given
+        (2*x**3 + 2*x**2*y + 6*x**2 + 2*x*y + 8*x + 4, 0)
+
+        See Also
+        ========
+
+        prem, pquo, pexquo
+
+        """
+        f = self
+        x = f.ring.index(x)
+
+        df = f.degree(x)
+        dg = g.degree(x)
+
+        if dg < 0:
+            raise ZeroDivisionError("polynomial division")
+
+        q, r, dr = x, f, df
+
+        if df < dg:
+            return q, r
+
+        N = df - dg + 1
+        lc_g = g.coeff_wrt(x, dg)
+
+        xp = f.ring.gens[x]
+
+        while True:
+
+            lc_r = r.coeff_wrt(x, dr)
+            j, N = dr - dg, N - 1
+
+            Q = q * lc_g
+
+            q = Q + (lc_r)*xp**j
+
+            R = r * lc_g
+
+            G = g * lc_r * xp**j
+
+            r = R - G
+
+            dr = r.degree(x)
+
+            if dr < dg:
+                break
+
+        c = lc_g**N
+
+        q = q * c
+        r = r * c
+
+        return q, r
+
+    def pquo(self, g, x=None):
+        """
+        Polynomial exact pseudo-quotient in multivariate polynomial ring.
+
+        Examples
+        ========
+        >>> from sympy.polys import ring, ZZ
+        >>> R, x,y = ring("x,y", ZZ)
+        >>> f = x**2 + x*y
+        >>> g = 2*x + 2*y
+        >>> h = 2*x + 2
+        >>> f.pquo(g)
+        2*x
+        >>> f.pquo(h)
+        2*x + 2*y - 2
+
+        See Also
+        ========
+
+        prem, pdiv, pexquo
+
+        """
+        f = self
+        x = f.ring.index(x)
+        return f.pdiv(g, x)[0]
```

Comment:
It is redundant to call index here when `pdiv` will do it again anyway. Same in pexquo